core/GlobalState.h

```diff
@@ -227,6 +240,12 @@ class GlobalState final {
     bool hasAnyDslPlugin() const;
 
     std::vector<std::unique_ptr<pipeline::semantic_extension::SemanticExtension>> semanticExtensions;
+    // In LSP mode: Used to pre-empt typechecking (post-resolver).
+    // - Worker threads grab as a reader lock, and routinely gives up and re-acquire the lock to allow other requests to
+    // pre-empt.
+    // - Typechecking coordinator thread grabs as a writer lock when there's a preemption function, which halts all
+    // worker threads.
+    const std::unique_ptr<absl::Mutex> typecheckMutex;
```

Comment:
This should probably be operated on via `Unfreeze*` classes rather than manually